src/shared/cgroup-setup.c

```diff
@@ -17,6 +17,67 @@
 #include "fileio.h"
 #include "user-util.h"
 #include "fd-util.h"
+#include "virt.h"
+
+static bool cg_is_unified_possible(void) {
+        _cleanup_free_ char *buf = NULL;
+        _cleanup_strv_free_ char **lines = NULL;
+        char **line;
+        int r;
+
+        r = read_full_virtual_file("/proc/cgroups", &buf, NULL);
+        if (r < 0) {
+                log_error_errno(r, "Could not read /proc/cgroups, ignoring.");
+                return true;
+        }
+
+        /* The intention of this is to check if the fully unified
+         * cgroup tree setup is possible, meaning all enabled kernel
+         * cgroup controllers are currently not in use by cgroup1.
+         * For reference:
+         * https://systemd.io/CGROUP_DELEGATION/#three-different-tree-setups-
+         *
+         * Note that this is typically only useful to check inside a
+         * container where we don't know what cgroup tree setup is in
+         * use by the host; if the host is using legacy or hybrid,
+         * we can't use unified since some or all controllers would
+         * be missing. This is not the best way to detect this, as
+         * whatever container manager created our container should
+         * have mounted /sys/fs/cgroup appropriately, but in case
+         * that wasn't done, we try to detect if it's possible for
+         * us to use unified cgroups. */
```

Comment:
we usually break our code at 109 chars btw, see .vimrc and .dir-locals.el

Reply:
ack, thanks